### PR TITLE
MM-27936: Allow resizing the channel sidebar and right-hand sidebar

### PR DESCRIPTION
**Pull request by kyeongsoosoo**
#### Summary

Previous WebApp PR: https://github.com/mattermost/mattermost-webapp/pull/11864

- make `lhs` and `rhs` resizable
- add `size` property to sidebar store
#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/21675

#### Screenshots

For window widths <900:

not resizable

For window widths 900-1199:

https://user-images.githubusercontent.com/59413880/209613520-4defe4d2-8cd4-47ac-9b3c-56929e8ed2a1.mov

For window widths 1200-1679:

https://user-images.githubusercontent.com/59413880/209613621-411d0279-c16b-4bdf-b430-4130b41d6d43.mov

For window widths >= 1680px:

https://user-images.githubusercontent.com/59413880/209613761-d9d2c49b-144f-4def-b784-1f4c8776ba0b.mov

#### Release Note

```release-note
Allow resizing the channel sidebar and right-hand sidebar
```

### CODE REVIEW

**Comment by kyeongsoosoo** [comment]
@nevyangelova I re-opened the PR and checked that it's working well :)

**Comment by matthewbirtch** [comment]
Is there an issue with spinning up test servers?

**Comment by AshishDhama** [comment]
@kyeongsoosoo could you please fix the failing test.

File: webapp/channels/src/stores/local_storage_store.ts
Comment (kyeongsoosoo) [reply webapp/channels/src/stores/local_storage_store.ts:19]: There was a review in the previous PR, but it seems to be missing.
@esethna Could you check it out?
It is about saving the resize state of the sidebar

스크린샷 2023-05-01 오후 11 37 52

**Comment by kyeongsoosoo** [comment]
> @kyeongsoosoo could you please fix the failing test.

@AshishDhama I think it was fixed by updating the branch

**Comment by kyeongsoosoo** [comment]
> @kyeongsoosoo could you please fix the failing test.

@AshishDhama 
Circleci test appears to be failing, but I don't have access to view it.
Could you tell me how to see it?

File: webapp/channels/src/utils/constants.tsx
Comment (AshishDhama) [inline webapp/channels/src/utils/constants.tsx:2039]: @kyeongsoosoo Could you please move these to separate file.

**Review by AshishDhama — CHANGES_REQUESTED** [review]

**Comment by AshishDhama** [comment]
> > @kyeongsoosoo could you please fix the failing test.
> 
> @AshishDhama Circleci test appears to be failing, but I don't have access to view it. Could you tell me how to see it?

I reran the test it passed.

**Comment by nevyangelova** [comment]
@kyeongsoosoo a fixed was merged to the mattermost server master branch, could you update your fork ? Thank you :)

File: webapp/channels/src/stores/local_storage_store.ts
Comment (matthewbirtch) [reply webapp/channels/src/stores/local_storage_store.ts:19]: I don't think we need to sync across all clients. As long as the resize state can be maintained on refresh from the same client, I think that would suffice.

Comment (kyeongsoosoo) [reply webapp/channels/src/stores/local_storage_store.ts]: I changed it not to use user_id.

File: webapp/channels/src/utils/constants.tsx
Comment (kyeongsoosoo) [reply webapp/channels/src/utils/constants.tsx]: I moved these to the sidebar component folder.

File: webapp/channels/src/stores/local_storage_store.ts
Comment (kyeongsoosoo) [reply webapp/channels/src/stores/local_storage_store.ts]: I changed it not to use user_id.

**Comment by matthewbirtch** [comment]
@AshishDhama any idea why the test server won't spin up?

File: webapp/channels/src/sass/responsive/_tablet.scss
Comment (kyeongsoosoo) [reply webapp/channels/src/sass/responsive/_tablet.scss]: I removed `important` by modifying component logic that was affecting css

**Comment by AshishDhama** [comment]
> @AshishDhama any idea why the test server won't spin up?

Mostly due to failing checks,

@kyeongsoosoo have you synced master branch of your fork with latest master.

**Comment by AshishDhama** [comment]
@kyeongsoosoo @matthewbirtch Test server is up

**Comment by kyeongsoosoo** [comment]
@matthewbirtch Fixed a problem that could occur when resizing browser width

**Review by matthewbirtch — APPROVED** [review]

File: webapp/channels/src/components/resizable_sidebar/resizable/index.tsx
Comment (AshishDhama) [inline webapp/channels/src/components/resizable_sidebar/resizable/index.tsx:191]: @kyeongsoosoo What is the benefit of this if statement?
cc// @M-ZubairAhmed @hmhealey

**Comment by AshishDhama** [comment]
@kyeongsoosoo Looking good just one question

**Review by AshishDhama — APPROVED** [review]

File: webapp/channels/src/components/resizable_sidebar/resizable/index.tsx
Comment (kyeongsoosoo) [reply webapp/channels/src/components/resizable_sidebar/resizable/index.tsx:191]: To prevent unnecessary event registration, the event was registered at the moment of pressing the resize line

**Comment by hmhealey** [comment]
I removed the stale label since this is waiting on reviews from our side

**Comment by M-ZubairAhmed** [comment]
@calebroseland gentle ping on this one 😄

**Comment by yasserfaraazkhan** [comment]
Hi @M-ZubairAhmed 
this change is not working on the spinwick, can you help in verifying ?

**Review by calebroseland — CHANGES_REQUESTED** [review]
Hi @kyeongsoosoo, I've attempted to review this a few times now, and I keep running into mental blocks of my own making. I think what I've been running against is a difference in approach and paradigm.

Your PR addresses quite a number of UX details, and your attention to detail there is fantastic. At the same time, there's quite a lot of code and infrastructure, and it's a very JS-oriented approach.

`Resizable` is pretty well-written and handles quite a few details, but there're still a lot of layout effects and other event hooks and default props—some absolutely necessary, for sure, but there are a good number of dependencies and conditionals—onLimitChange, min/max, initialWidth vs. defaultWidth, etc. There's also unnecessary memoization in a number of places over in `ResizableLhs`/`ResizableRhs`

When we overhauled the main layout last year (https://github.com/mattermost/mattermost-webapp/pull/11019), we used a very CSS-based paradigm and approach to the main layout and responsiveness.

Back to resizing: traditionally or conventionally, directly overriding an element's `style.width` property is often used. For simple cases, this is absolutely fine, but one big caveat is that is can bypass any number of CSS rules in the process—that can be a problem in intricate scenarios, sometimes requiring emulating responsive and other CSS-based logic in JS (possible, but hard to do).

#### Proposal
I wanted to submit the idea of pivoting a bit to use simpler infrastructure, with a CSS-based approach using CSS Vars to provide an override value to the same CSS rules that normally define the respective elements `width`s. 

This does a few things IMHO:

1. Layout, styling, and min/max logic, and responsiveness all (generally) stay within CSS, where they fit most naturally.
2. Event listeners, preference storage, and other misc things generally stay within JS.

As an example, https://github.com/mattermost/mattermost/pull/23736 may explain better. It's 
based on parts of your PR. It's certainly not complete, and missing some details that are included in your PR, but it does condense a decent amount of the infrastructure into their respective areas (fewer files changed). 

Please feel free to take this and do as you see fit with it. Let me know what you think, and thanks again for all your work on this!

**Comment by calebroseland** [comment]
Also found some quirks/edge cases while testing:

https://github.com/mattermost/mattermost/assets/11724372/8d196378-c2b0-4254-a811-00eb1e01e1b5

**Comment by kyeongsoosoo** [comment]
@calebroseland Thanks for the review. I agree that the js-oriented approach has a caveat.
I think it'd be good to approach it with a css-based way. I'll try to apply the infrastructure.

**Comment by johndavidlugtu** [comment]
@kyeongsoosoo friendly ping here -- let us know if you have some questions for the team

**Comment by kyeongsoosoo** [comment]
@calebroseland 
I changed the code to use a css-based approach.
I think it was a good choice as the code got cleaner.
I also fixed other reviews, including fixing edge cases.
Sorry for the delay🥲

**Comment by matthewbirtch** [comment]
I'll review this again once we get a test server up and running. Seems like it's not generating ATM

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Took another look at this and it's working great. A couple of minor things I noticed:

1. When double-clicking to reset the RHS width to the default, can we remove the animation and just make it immediately go to the default width without any transition?
2. For browser window widths between 769-900px, I believe we spec'd it to not be resizable at all, but we should allow resizing to the minimum of 200px and max 264px still I think at these widths. I can update the spec to reflect this
3. There is something a little odd happening when opening the RHS. While on a large screen, I resized the RHS to it's max-width. Then I closed the RHS and reduced the window size to a smaller width. Up re-opening the RHS it first opens wide and after a short delay collapses its width to the proper smaller width (see video below).

https://github.com/mattermost/mattermost/assets/2040554/0dc02a61-5192-42bf-a7eb-fd5ec9c7c613

**Comment by kyeongsoosoo** [comment]
> Took another look at this and it's working great. A couple of minor things I noticed:

@matthewbirtch I've updated based on the reviews☺️

**Review by matthewbirtch — APPROVED** [review]
Awesome stuff @kyeongsoosoo!

**Comment by harshilsharma63** [comment]
Damn! I love this PR 😍😍😍😍 Thanks for making this happen!

**Review by harshilsharma63 — APPROVED** [review]

**Comment by harshilsharma63** [comment]
I've fixed the part where the widths were scoped to a single team. Now every team shares the same width.

**Comment by kyeongsoosoo** [comment]
> I've fixed the part where the widths were scoped to a single team. Now every team shares the same width.

Thanks for the review😄

**Review by yasserfaraazkhan — APPROVED** [review]

File: webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx
Comment (calebroseland) [inline webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx]: this was a selector function, not a string, so incompatible as-used; removed as suffixed by 3rd param/currentUserID

**Review by calebroseland — APPROVED** [review]
I still question the need for the extra event listeners and new/separate breakpoints, but this is getting to be good enough and it could possibly be iterated on in the future.

I was seeing some responsive issues with the LHS header—at it's narrowest it would cut off the `+` menu, so I pushed a change to address that:

| before | after |
|---|---|
|  CleanShot 2023-08-24 at 12 07 14 |  CleanShot 2023-08-28 at 17 37 23 |


There's still one odd UX bit—not critical IMO, but at viewport widths less than `900` wide (but still over the mobile threshold of `768`), the resize handle doesn't work (but still grab-able) and it snaps back to its default size if it was previously resized.


https://github.com/mattermost/mattermost/assets/11724372/e921e54c-35e2-4276-bc62-648f3f012d43

 Nonetheless, this is a fantastic piece of work and will be really nice to get this in! 
 
 Thanks, @kyeongsoosoo!

File: webapp/channels/src/components/advanced_text_editor/formatting_bar/hooks.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/advanced_text_editor/formatting_bar/hooks.tsx:38]: [ref's should not](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/) be used as dependency

File: webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx:239]: nit : return null when disabled

File: webapp/channels/src/components/window_size_observer/WindowSizeObserver.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/window_size_observer/WindowSizeObserver.tsx:90]: nit can we return null instead of empty jsx?

**Review by M-ZubairAhmed — COMMENTED** [review]
Looks and works great. Just some nits

File: webapp/channels/src/components/advanced_text_editor/formatting_bar/hooks.tsx
Comment (kyeongsoosoo) [reply webapp/channels/src/components/advanced_text_editor/formatting_bar/hooks.tsx]: Thanks for the link. I fixed it.

File: webapp/channels/src/components/window_size_observer/WindowSizeObserver.tsx
Comment (kyeongsoosoo) [reply webapp/channels/src/components/window_size_observer/WindowSizeObserver.tsx]: I changed it to return null.

File: webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx
Comment (kyeongsoosoo) [reply webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx:239]: Done!

**Review by M-ZubairAhmed — APPROVED** [review]
This looks good :shipit: Thank you

**Comment by marianunez** [comment]
@kyeongsoosoo thank you for this great contribution and your patience working through the feedback! This is a high-impact feature 🚀

**Pull request merged**
